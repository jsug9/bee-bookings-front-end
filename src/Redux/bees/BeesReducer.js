/* eslint-disable no-param-reassign */
import axios from 'axios';
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const GET_BEES = 'Bees/GET_BEES';
const GET_BEE_DETAILS = 'Bees/GET_BEE_DETAILS';
const ADD_BEE = 'Bees/ADD_BEE';
const DELETE_BEE = 'Bees/DELETE_BEE';

const getBeesEndpoint = 'https://bee-store.herokuapp.com/api/v1/items';

const initialState = {
  allBees: [],
  beeDetails: {},
  isLoading: false,
};

export const getAllBees = createAsyncThunk(GET_BEES, async () => {
  const { data } = await axios.get(getBeesEndpoint);
  return data;
});

export const getBeeDetails = createAsyncThunk(GET_BEE_DETAILS, async (id) => {
  const { data } = await axios.get(`${getBeesEndpoint}/${id}`);
  return data;
});

export const addBee = createAsyncThunk(ADD_BEE, async (bee) => {
  axios.post(getBeesEndpoint, bee);
  const { data } = await axios.get(getBeesEndpoint);
  return data;
});

export const deleteBee = createAsyncThunk(DELETE_BEE, async (beeId) => {
  axios.delete(`${getBeesEndpoint}/${beeId}`);
  const { data } = await axios.get(getBeesEndpoint);
  return data;
});

const beesSlice = createSlice({
  name: 'Bees',
  initialState,
  reducers: {},
  extraReducers: {
    [getAllBees.fulfilled]: (state, action) => {
      state.allBees = action.payload.sort((a, b) => a.id - b.id);
      state.isLoading = false;
    },
    [getAllBees.rejected]: (state) => {
      state.allBees = [];
      state.isLoading = false;
    },
    [getAllBees.pending]: (state) => {
      state.allBees = [];
      state.isLoading = true;
    },
    [getBeeDetails.fulfilled]: (state, action) => {
      state.beeDetails = action.payload;
      state.isLoading = false;
    },
    [getBeeDetails.rejected]: (state) => {
      state.beeDetails = {};
      state.isLoading = false;
    },
    [getBeeDetails.pending]: (state) => {
      state.beeDetails = {};
      state.isLoading = true;
    },
    [addBee.fulfilled]: (state, action) => {
      state.allBees = action.payload;
      state.isLoading = false;
    },
    [addBee.rejected]: (state) => {
      state.allBees = [];
      state.isLoading = false;
    },
    [addBee.pending]: (state) => {
      state.allBees = [];
      state.isLoading = true;
    },
    [deleteBee.fulfilled]: (state, action) => {
      state.allBees = action.payload;
      state.isLoading = false;
    },
    [deleteBee.rejected]: (state) => {
      state.allBees = [];
      state.isLoading = false;
    },
    [deleteBee.pending]: (state) => {
      state.allBees = [];
      state.isLoading = true;
    },
  },
});

export default beesSlice.reducer;
