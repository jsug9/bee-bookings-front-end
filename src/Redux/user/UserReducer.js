/* eslint-disable no-param-reassign */
import axios from 'axios';
import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

const signInEndpoint = 'https://bee-store.herokuapp.com/api/v1/signin';
const signUpEndpoint = 'https://bee-store.herokuapp.com/api/v1/signup';

export const signIn = createAsyncThunk('user/signIn', async (username) => {
  const userParams = { user: { username } };
  const { data } = await axios.post(signInEndpoint, userParams);
  return data;
});

export const sum = () => 'from user reducer';

export const signUp = createAsyncThunk('user/signUp', async (username) => {
  const userParams = { user: { username } };
  const { data } = await axios.post(signUpEndpoint, userParams);
  return data;
});

const initialState = {
  username: null,
  userId: null,
  isLoading: false,
};

const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    logOutUser: (state) => {
      state.username = null;
      state.userId = null;
      state.createdUser = null;
    },
  },
  extraReducers: {
    [signIn.pending]: (state) => {
      state.isLoading = true;
    },
    [signIn.fulfilled]: (state, action) => {
      state.username = action.payload.username;
      state.userId = action.payload.user_id;
      state.isLoading = false;
    },
    [signIn.rejected]: (state) => {
      state.isLoading = false;
    },
    [signUp.pending]: (state) => {
      state.isLoading = true;
    },
    [signUp.fulfilled]: (state, action) => {
      state.username = action.payload.username;
      state.userId = action.payload.user_id;
      state.isLoading = false;
      state.createdUser = true;
    },
    [signUp.rejected]: (state) => {
      state.isLoading = false;
    },
  },
});

export const { logOutUser } = userSlice.actions;
export default userSlice.reducer;
