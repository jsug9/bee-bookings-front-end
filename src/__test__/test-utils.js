/* eslint-disable react/prop-types */
import { BrowserRouter as Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import setupStore from '../Redux/testStore';

function renderWithProviders(
  ui,
  {
    preloadedState,
    store = setupStore(preloadedState),
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Router><Provider store={store}>{children}</Provider></Router>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from '@testing-library/react';

export default renderWithProviders;
