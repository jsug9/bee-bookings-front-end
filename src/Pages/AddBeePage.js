import { Button, TextField } from '@mui/material';
import LibraryAddIcon from '@mui/icons-material/LibraryAdd';
import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import UploadField from '../Components/UploadImage';
import UploadImage, { S3_BUCKET } from '../Utilities/AmazonUpload';
import { addBee } from '../Redux/bees/BeesReducer';
import '../Styles/BeeForm.scss';
import PleaseLogin from '../Components/PleaseLogin';

const AddBeePage = () => {
  const [selectedFile, setSelectedFile] = useState(null);
  const [progress, setProgress] = useState(0);
  const userId = useSelector((state) => state.user.userId);
  const dispatch = useDispatch();

  const handleSubmit = (e) => {
    e.preventDefault();
    const beeName = e.target.name.value;
    const beeDescription = e.target.description.value;
    const beeImage = `https://s3.amazonaws.com/${S3_BUCKET}/${selectedFile.name}`;

    if (beeName.trim() && beeDescription.trim() && beeImage) {
      const newBee = {
        name: beeName,
        description: beeDescription,
        image: beeImage,
      };
      UploadImage(selectedFile, setProgress, e, dispatch(addBee(newBee)));
    }
  };

  return (
    <section className="form-container">
      {userId
      && (
      <form className="bee-form" onSubmit={handleSubmit}>
        <h1>Add Your own Bee</h1>
        <div>
          <TextField
            required
            id="name-form"
            label="Name"
            name="name"
            style={{ backgroundColor: 'white', borderRadius: '4px' }}
          />
        </div>
        <div>
          <TextField
            required
            id="description-form"
            label="Description"
            multiline
            rows={4}
            name="description"
            style={{ backgroundColor: 'white', borderRadius: '4px' }}
          />
        </div>
        <div>
          <UploadField setSelectedFile={setSelectedFile} progress={progress} />
        </div>
        <Button
          variant="contained"
          color="success"
          startIcon={<LibraryAddIcon />}
          sx={{ fontWeight: 'bold' }}
          type="submit"
          className="submit-button"
        >
          Add bee
        </Button>
      </form>
      )}
      {!userId
      && (
        <PleaseLogin message="Add a Bee" />
      )}
    </section>
  );
};

export default AddBeePage;
