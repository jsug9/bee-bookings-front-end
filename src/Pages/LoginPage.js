import '../Styles/login_page.scss';
import { TextField } from '@mui/material';
import { useDispatch, useSelector } from 'react-redux';
import React, { useState } from 'react';
import { toggleSignup } from './SignupPage';
import { signIn } from '../Redux/user/UserReducer';

// Toggles visibility of the login page overlay
// NO TOUCHY: specially Augusto
export const toggleLogin = () => {
  const loginPage = document.getElementById('loginPage');
  loginPage.classList.toggle('invisible');
  loginPage.classList.toggle('flex');
};

// Toggles visibility of the login page overlay when the user clicks the sign up link
// NO TOUCHY: specially Augusto
const signupToggle = () => {
  toggleLogin();
  toggleSignup();
};

const LoginPage = () => {
  const [username, setUsername] = useState('');
  const user = useSelector((state) => state.user);

  const dispatch = useDispatch();
  const handleSubmit = (e) => {
    e.preventDefault();
    dispatch(signIn(username));
  };
  return (
    <div id="loginPage" className={user.userId ? 'hide' : 'invisible'}>
      <div id="loginClose" role="presentation" onClick={toggleLogin}>X</div>
      <div className="loginForm mobile">
        <h1>Buzz In</h1>
        <section className="loginSection" id="loginSection">
          <form className="loginForm" method="post" onSubmit={handleSubmit}>
            <TextField
              style={{ backgroundColor: 'white', borderRadius: '4px' }}
              type="text"
              name="username"
              id="username"
              label="Enter your username"
              required
              maxLength="30"
              value={username}
              onChange={(e) => setUsername(e.target.value)}
            />
            <span className="error" aria-live="polite" />
            <button type="submit" id="submitLogin" className="loginBtn">Let&apos;s Go!</button>
          </form>
          <div className="signupLink">
            <p id="signup" role="presentation" onClick={signupToggle}>Sign Up</p>
          </div>
        </section>
      </div>
    </div>
  );
};

export default LoginPage;
