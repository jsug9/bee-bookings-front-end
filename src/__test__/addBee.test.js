import { fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import LoginPage from '../Pages/LoginPage';
import AddBeePage from '../Pages/AddBeePage';
import renderWithProviders, { screen } from './test-utils';
import { signIn } from '../Redux/user/UserReducer';
import { getAllBees, addBee } from '../Redux/bees/BeesReducer';
import server from '../mswMocks/server';
import realStore from '../Redux/configureStore';

// Establish API mocking before all tests.
beforeAll(() => server.listen());
// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers());
// Clean up after the tests are finished.
afterAll(() => server.close());

it('Doesnt render the add bee page if the user is not logged in', async () => {
  const store = realStore;
  await store.dispatch(signIn('AaronIsCool'));
  renderWithProviders(<AddBeePage />, { store });
  expect(screen.queryByText(/Add Bee/)).not.toBeInTheDocument();
});

it('It renders the Add Bee Component if User is Signed In', async () => {
  const store = realStore;
  renderWithProviders(<LoginPage />, { store });

  expect(screen.getByText(/Sign Up/)).toBeInTheDocument();
  const input = document.querySelector('#username');
  const submitButton = screen.getByText(/Let's Go!/);
  expect(input).toBeInTheDocument();
  expect(submitButton).toBeInTheDocument();
  await act(async () => {
    input.value = 'AaronIsCool';
    fireEvent.submit(submitButton);
  });
  renderWithProviders(<AddBeePage />, { store });
  expect(screen.getByText(/Add Your own Bee/)).toBeInTheDocument();
});

it('Creates a bee when the user submits the form', async () => {
  const store = realStore;
  renderWithProviders(<LoginPage />, { store });

  expect(screen.getByText(/Sign Up/)).toBeInTheDocument();
  const input = document.querySelector('#username');
  const submitButton = screen.getByText(/Let's Go!/);
  expect(input).toBeInTheDocument();
  expect(submitButton).toBeInTheDocument();
  await act(async () => {
    input.value = 'AaronIsCool';
    fireEvent.submit(submitButton);
  });
  renderWithProviders(<AddBeePage />, { store });
  expect(screen.getByText(/Add Your own Bee/)).toBeInTheDocument();
  const form = document.querySelector('form');
  expect(form).toBeInTheDocument();
  document.querySelector('#name-form').value = 'Wonderful Bee';
  const description = document.querySelector('#description-form');
  description.value = 'This is a wonderful bee';
  expect(description.value).toBe('This is a wonderful bee');
  const fileinput = document.querySelector(
    '.MuiInputBase-input.MuiOutlinedInput-input.css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input',
  );
  fileinput.value = 'https://images.unsplash.com/photo-1518791841217-8f162f1e1131?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=800&q=60';
  await store.dispatch(getAllBees());
  expect(store.getState().bees.allBees.length).toBe(10);
  await store.dispatch(
    addBee({
      name: 'Wonderful Bee',
      description: 'This is a wonderful bee',
      image:
        'https://images.unsplash.com/photo-1518791841217-8f162f1e1131?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=800&q=60',
    }),
  );
  expect(store.getState().bees.allBees.length).toBe(10);
});

it('takes a snapshot of the addBee component', () => {
  const { asFragment } = renderWithProviders(<AddBeePage />);
  expect(asFragment()).toMatchSnapshot();
});
