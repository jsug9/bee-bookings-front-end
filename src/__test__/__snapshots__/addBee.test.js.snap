// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`takes a snapshot of the addBee component 1`] = `
<DocumentFragment>
  <section
    class="form-container"
  >
    <div>
      <h1
        class="please-log-in"
      >
        Please log in to add a bee
      </h1>
    </div>
  </section>
</DocumentFragment>
`;
